pipeline/common/datasets.py

```diff
@@ -90,7 +90,12 @@ def shuffle_with_max_lines(
         if len(lines) == max_lines:
             break
 
-    random.shuffle(lines)
+    # random.shuffle requires random access via indexing
+    # deque supports fast adding/removing from its ends with O(1)
+    # but indexing is O(N) which is too slow for shuffling large arrays
+    lines_list = list(lines)
```

Comment:
Thought: Ah, thanks for fixing this. It looks like we'll need multiple copies of the list in memory to do it this way, and iterate over the list again to do it. This will definitely fix the performance. If we want to reduce the memory usage back down and reduce the amount of iterations then we could do it by using a list instead of a deque, and then do some index counting to move through it. It may be worth filing as a perf & cost labeled bug, but this will fix it for now.